### PR TITLE
Add fuzz test for default path normalizer

### PR DESCRIPTION
**Pull request by yanavlasov**
Additional Description:
Fuzz test for path normalizer with some sanity checks.

Risk Level: Low
Testing: New fuzz test
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

Signed-off-by: Yan Avlasov <yavlasov@google.com>

### CODE REVIEW

File: test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz.proto [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+package test.extensions.http.header_validators.envoy_default;
+
+import "google/protobuf/empty.proto";
+import "validate/validate.proto";
+
+import "envoy/extensions/http/header_validators/envoy_default/v3/header_validator.proto";
+
+message PathNormalizerFuzzTestCase {
+  // the character set of the :path and :method headers is validated before normalization.
```
Comment (KBaichoo) [inline test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz.proto:11]: s/the/The/

File: test/extensions/http/header_validators/envoy_default/path_normalizer_corpus/relatrive_with_percent_2f [hunk 2]
```diff
@@ -0,0 +1,9 @@
+method: "POST"
```
Comment (KBaichoo) [inline test/extensions/http/header_validators/envoy_default/path_normalizer_corpus/relatrive_with_percent_2f:1]: rename file s/relatrive_with_percent_2f/relative_with_percent_2f/

File: test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz.proto [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+  // As such it is ok to use strings to represent these values in the fuzz test.
+  string method = 1;
+  string path = 2;
+
+  envoy.extensions.http.header_validators.envoy_default.v3.HeaderValidatorConfig config = 3;
```
Comment (KBaichoo) [inline test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz.proto:16]: What's the reason we opted for the encapping message vs UriPathNormalizationOptions? The latter seems to better target the space of this fuzzer.

File: test/extensions/http/header_validators/envoy_default/path_normalizer_corpus/corpus-1 [hunk 4]
```diff
@@ -0,0 +1,9 @@
+method: "GET"
```
Comment (KBaichoo) [inline test/extensions/http/header_validators/envoy_default/path_normalizer_corpus/corpus-1:1]: Nice job naming the other files, they really capture the test case. Can this have a similar descriptive name?

File: test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz_test.cc [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+  method.setCopyUnvalidatedForTestOnly(input.method());
+  path.setCopyUnvalidatedForTestOnly(input.path());
+
+  NiceMock<Envoy::StreamInfo::MockStreamInfo> stream_info;
+  Extensions::Http::HeaderValidators::EnvoyDefault::Http1HeaderValidator validator(
+      input.config(), Http::Protocol::Http11, stream_info);
+  // The character set of the :path and :method headers is validated before normalization.
+  // Here we will just not run the test with invalid values
+  if (!validator.validateMethodHeader(method) ||
+      (!(method == "CONNECT" && path.empty()) && !validator.validatePathHeaderCharacters(path))) {
```
Comment (KBaichoo) [inline test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz_test.cc:26]: I think `(!(method == "CONNECT" && path.empty()) && !validator.validatePathHeaderCharacters(path)))`
should be 
`(!(method == "CONNECT" && path.empty()) || !validator.validatePathHeaderCharacters(path)))`

Since the validations seem orthogonal

**Review by KBaichoo — COMMENTED** [review]
Thank you for adding this! Mostly looks good.

File: test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz.proto (hunk 1 above)
Comment (yanavlasov) [reply test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz.proto]: Done

File: test/extensions/http/header_validators/envoy_default/path_normalizer_corpus/relatrive_with_percent_2f (hunk 2 above)
Comment (yanavlasov) [reply test/extensions/http/header_validators/envoy_default/path_normalizer_corpus/relatrive_with_percent_2f]: Done

File: test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz.proto (hunk 3 above)
Comment (yanavlasov) [reply test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz.proto]: You are correct. Fixed.

File: test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz_test.cc (hunk 5 above)
Comment (yanavlasov) [reply test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz_test.cc:28]: I only want skip calling `validatePathHeaderCharacters` in case method is `CONNECT` and path is empty. In this case the && should be correct. If I change it || the `validatePathHeaderCharacters`  will still be called with empty path for CONNECT method.

Comment (KBaichoo) [reply test/extensions/http/header_validators/envoy_default/path_normalizer_fuzz_test.cc:28]: sg, thanks for the explaination

**Review by KBaichoo — APPROVED** [review]
Thank you for adding this fuzzer!
